loom/common/platform/platformNetwork.c

```diff
@@ -399,80 +476,256 @@ void loom_net_readTCPSocket(loom_socketId_t s, void *buffer, int *bytesToRead, i
+            // How many bytes we are going to send this time
+            bytesSending = bytesPending < writeChunkMaxSize ? bytesPending : writeChunkMaxSize;
+
+            // How many bytes we can actually read at once from the buffer
+            // at this time.
+            bytesAvailable = bipbuf_available(bipbuf);
+            if (bytesAvailable < bytesSending) bytesSending = bytesAvailable;
+
+            // Pointer into the buffer
+            buf = bipbuf_peek(bipbuf, bytesSending);
+            lmAssert(buf, "Unable to peek into bipbuffer, %d pending, %d sending, %d available", bytesPending, bytesSending, bytesAvailable);
 
-                // Set up to try again by advancing into the buffer.
-                buffer = (void *)((char *)buffer + result);
-                continue;
+            result = send((SOCKET)(size_t)socket, buf, bytesSending, MSG_NOSIGNAL);
+
+            if (result >= 0)
+            {
+                lmAssert(bipbuf_poll(bipbuf, result), "Internal error, unable to poll the bipbuffer");
+                sent += result;
+                // Uncomment to print out every successful write
+                //platform_debugOut("Pumped %d bytes", result);
+                if (bipbuf_used(bipbuf) > 0)
+                {
+                    // Uncomment to print out partial writes
+                    //platform_debugOut("Partial write %d sent, %d left,   %d sent this call", result, bipbuf_used(bipbuf), sent);
+                    continue;
+                }
+                break;
             }
 
-            return bytesToWrite - bytesLeft;
+            // Try another time on failed write
+            break;
         }
 
-#if LOOM_PLATFORM == LOOM_PLATFORM_WIN32
-        winsockErrorCode = WSAGetLastError();
-        if (winsockErrorCode != WSAEWOULDBLOCK)
+        sd->bytesSent += sent;
+
+        // Stall check
+        if (sent == 0)
         {
-            break;
+            sd->stallWrites++;
+            if (sd->stallWrites >= 200)
+            {
+                if (sd->stallWrites == 200)
+                {
+                    loom_resetTimer(sd->activityTimer);
+                }
+                else
+                {
+                    if (loom_readTimer(sd->activityTimer) > 30000)
+                    {
+                        sd->stalled = 1;
+                    }
+                }
+            }
+            // Uncomment to print out stalled write status
+            //platform_debugOut("Stall check: %d writes, %d stalled, %d ms", sd->stallWrites, sd->stalled, loom_readTimer(sd->activityTimer));
         }
-#else
-        if (errno != EAGAIN)
+        else
         {
-            break;
+            sd->stallWrites = 0;
+            sd->stalled = 0;
         }
-#endif
 
-        if (loom_net_isSocketDead(s))
+        // Debug print log output
+        if (print) {
+            platform_debugOut("Socket %x%s write buffer status: %d %s/s, %lld%% full, %d left, %d free, %d total",
```

Comment:
Not without additional work on the logging system. All the log messages get sent through a socket, so if you try to send logging logs of logging while logging you'll have a bad time.